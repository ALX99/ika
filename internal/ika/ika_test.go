package ika

import (
	"context"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"path"
	"testing"
	"time"

	"github.com/alx99/ika"
	"github.com/alx99/ika/internal/config"
	iplugins "github.com/alx99/ika/internal/plugins"
	"github.com/alx99/ika/plugins"
	"github.com/gkampitakis/go-snaps/match"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/matryer/is"
)

func TestMain(m *testing.M) {
	v := m.Run()
	snaps.Clean(m)
	os.Exit(v)
}

func TestAnyMethod(t *testing.T) {
	t.Parallel()
	is := is.New(t)
	c, baseURL := runServer(t)

	methods := []string{"GET", "POST", "PUT", "PATCH", "DELETE", "CONNECT", "OPTIONS", "TRACE"}
	for _, method := range methods {
		for _, req := range makeReqs(t, method, "/any", nil, baseURL, "testns1", "testns1.com") {
			resp, err := c.Do(req)
			is.NoErr(err)
			snapshotBody(t, resp)
		}
	}
}

func TestQueryParamsArePassedCorrectly(t *testing.T) {
	t.Parallel()
	is := is.New(t)
	c, baseURL := runServer(t)

	testCases := []struct {
		url      string
		expected map[string]string
	}{
		{url: "/get?hi=1"},
		{url: "/get?hi=1&bye=2"},
		{url: "/get?hi=%20hello%20world%20"},
		{url: "/get"},
		{url: "/get?hi=1&hi=2&hi=3"},
		{url: "/get?hi=&bye="},
		{url: "/get?hi=null"},
		{url: "/get?hi=true&bye=false"},
		{url: "/get?hi=123&bye=456.789"},
		{url: "/get?hi=1&bye=true&foo=null&bar=%20space%20"},
		{url: "/get?hi=hello%20world&bye=goodbye%2Fworld"},
	}

	for _, tc := range testCases {
		for _, req := range makeReqs(t, "GET", tc.url, nil, baseURL, "testns1", "testns1.com") {
			resp, err := c.Do(req)
			is.NoErr(err)
			snapshotBody(t, resp)
		}
	}
}

func TestWildcardRewrite(t *testing.T) {
	t.Parallel()
	is := is.New(t)
	c, baseURL := runServer(t)

	urls := []string{
		"/httpbun/any/a/huhh",
		"/httpbun/any/a/huhh?abc=lol&x=b",
		"/httpbun/any/a/huhh?abc=魚&x=は",
		"/httpbun/any/slash%2Fshould-bekept/next",
	}

	for _, url := range urls {
		for _, req := range makeReqs(t, "GET", url, nil, baseURL, "testns1", "testns1.com") {
			resp, err := c.Do(req)
			is.NoErr(err)
			snapshotBody(t, resp)
		}
	}
}

func TestRetainHostHeader(t *testing.T) {
	t.Parallel()
	is := is.New(t)
	c, baseURL := runServer(t)

	for _, req := range makeReqs(t, "GET", "/retain-host", nil, baseURL, "testns1", "testns1.com") {
		resp, err := c.Do(req)
		is.NoErr(err)
		snapshotBody(t, resp)
	}
}

func TestPathRewrite(t *testing.T) {
	t.Parallel()
	is := is.New(t)
	c, baseURL := runServer(t)

	urls := []string{
		"/path-rewrite/a/efgh",
		"/path-rewrite/a/huhh",
		"/path-rewrite/a/huhh?abc=lol&x=b",
		"/path-rewrite/a/huhh?abc=魚&x=は",
		"/path-rewrite/slash%2Fshould-bekept/next",
	}

	for _, url := range urls {
		for _, req := range makeReqs(t, "GET", url, nil, baseURL, "testns1", "testns1.com") {
			resp, err := c.Do(req)
			is.NoErr(err)
			snapshotBody(t, resp)
		}
	}
}

func TestRedirectsNotFollowed(t *testing.T) {
	t.Parallel()
	is := is.New(t)
	c, baseURL := runServer(t)

	// don't redirect
	c.CheckRedirect = func(req *http.Request, via []*http.Request) error {
		return http.ErrUseLastResponse
	}

	for _, req := range makeReqs(t, "GET", "/httpbun/redirect-to?url=https%3A%2F%2Fgoogle.com", nil, baseURL, "testns1", "testns1.com") {
		resp, err := c.Do(req)
		is.NoErr(err)
		snaps.MatchSnapshot(t, stripVariableData(t, resp))
	}
}

func TestOnlyGetAllowed(t *testing.T) {
	t.Parallel()
	is := is.New(t)
	c, baseURL := runServer(t)

	getUrls := []string{
		"/only-get",
	}

	for _, url := range getUrls {
		for _, req := range makeReqs(t, "GET", url, nil, baseURL, "testns1", "testns1.com") {
			resp, err := c.Do(req)
			is.NoErr(err)
			snapshotBody(t, resp)
		}
	}

	postUrls := []string{
		"/only-get",
	}

	for _, url := range postUrls {
		for _, req := range makeReqs(t, "POST", url, nil, baseURL, "testns1", "testns1.com") {
			resp, err := c.Do(req)
			is.NoErr(err)
			snapshotBody(t, resp)
		}
	}
}

func TestNonTerminatedPaths(t *testing.T) {
	t.Parallel()
	is := is.New(t)
	c, baseURL := runServer(t)

	urls := []string{
		"/not-terminated/hi/",
		"/not-terminated/a/b/c/",
		"/not-terminated/a/b/c/d",
	}

	for _, url := range urls {
		for _, req := range makeReqs(t, "GET", url, nil, baseURL, "testns1", "testns1.com") {
			resp, err := c.Do(req)
			is.NoErr(err)
			snapshotBody(t, resp)
		}
	}
}

func TestTerminatedPaths(t *testing.T) {
	t.Parallel()
	is := is.New(t)
	c, baseURL := runServer(t)

	validUrls := []string{
		"/terminated/hi/",
	}

	for _, url := range validUrls {
		for _, req := range makeReqs(t, "GET", url, nil, baseURL, "testns1", "testns1.com") {
			resp, err := c.Do(req)
			is.NoErr(err)
			snapshotBody(t, resp)
		}
	}

	invalidUrls := []string{
		baseURL + "/terminated/hi/a/b/c/",
	}

	for _, url := range invalidUrls {
		for _, req := range makeReqs(t, "GET", url, nil, baseURL, "testns1", "testns1.com") {
			resp, err := c.Do(req)
			is.NoErr(err)
			is.Equal(resp.StatusCode, http.StatusNotFound)
			snapshotBody(t, resp)
		}
	}
}

func TestPassthrough(t *testing.T) {
	t.Parallel()
	is := is.New(t)
	c, baseURL := runServer(t)

	urls := []string{
		baseURL + "/passthrough/get",
		baseURL + "/passthrough/any/hihi/%2F",
	}

	for _, url := range urls {
		resp, err := c.Get(url)
		is.NoErr(err)

		snapshotBody(t, resp)
	}

	invalidUrls := []string{
		baseURL + "/passthrough%2F",
	}

	for _, url := range invalidUrls {
		req, err := http.NewRequest("GET", url, nil)
		is.NoErr(err)
		req.Host = "passthrough.com"

		resp, err := c.Do(req)
		is.NoErr(err)
		is.Equal(resp.StatusCode, http.StatusNotFound)
	}
}

func snapshotBody(t *testing.T, resp *http.Response) {
	t.Helper()
	is := is.New(t)
	body, err := io.ReadAll(resp.Body)
	is.NoErr(err)

	if resp.Header.Get("Content-Type") == "application/json" {
		snaps.MatchJSON(t, body, match.Any("headers.Referer").ErrOnMissingPath(false))
	} else {
		snaps.MatchSnapshot(t, body)
	}
}

func makeReqs(t *testing.T, method, url string, body io.Reader, baseURL, ns, host string) []*http.Request {
	t.Helper()
	is := is.New(t)
	var reqs []*http.Request

	if ns != "" {
		req, err := http.NewRequest(method, baseURL+"/"+path.Join(ns, url), body)
		is.NoErr(err)
		reqs = append(reqs, req)
	}

	if host != "" {
		req, err := http.NewRequest(method, url, body)
		is.NoErr(err)
		req.Host = host
	}
	return reqs
}

func stripVariableData(t *testing.T, resp *http.Response) *http.Response {
	t.Helper()
	resp.Request.Host = "<replaced>"
	resp.Request.URL.Host = "<replaced>"
	resp.Header.Del("Date")
	resp.Header.Del("X-Powered-By")
	return resp
}

func runServer(t *testing.T) (*http.Client, string) {
	t.Helper()
	is := is.New(t)
	s := newTestServer(t, httptest.NewUnstartedServer(nil))
	ctx, cancel := context.WithCancel(context.Background())
	t.Cleanup(cancel)
	go func() {
		time.Sleep(10 * time.Second)
		cancel()
	}()

	cfg, err := readConfig()
	is.NoErr(err)

	opts := config.Options{
		Plugins: map[string]ika.PluginFactory{
			"basic-modifier": plugins.ReqModifier{},
			"accessLog":      plugins.AccessLogger{},
			"dumper":         iplugins.Dumper{},
		},
	}

	errCh := make(chan error, 1)
	go func() {
		flush, err := run(ctx, newMakeTestServer(t, s), cfg, opts)
		flush()
		if err != nil {
			errCh <- err
		}
	}()

	select {
	case err := <-errCh:
		is.NoErr(err)
	case <-s.startCh:
	}

	return s.server.Client(), s.server.URL
}
