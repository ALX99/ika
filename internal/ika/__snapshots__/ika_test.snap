
[TestQueryParamsArePassedCorrectly - 1]
{
 "args": {
  "hi": "1"
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any?hi=1"
}
---

[TestQueryParamsArePassedCorrectly - 2]
{
 "args": {
  "bye": "2",
  "hi": "1"
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any?hi=1&bye=2"
}
---

[TestQueryParamsArePassedCorrectly - 3]
{
 "args": {
  "hi": " hello world "
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any?hi=%20hello%20world%20"
}
---

[TestQueryParamsArePassedCorrectly - 4]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestQueryParamsArePassedCorrectly - 5]
{
 "args": {
  "hi": [
   "1",
   "2",
   "3"
  ]
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any?hi=1&hi=2&hi=3"
}
---

[TestQueryParamsArePassedCorrectly - 6]
{
 "args": {
  "bye": "",
  "hi": ""
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any?hi=&bye="
}
---

[TestQueryParamsArePassedCorrectly - 7]
{
 "args": {
  "hi": "null"
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any?hi=null"
}
---

[TestQueryParamsArePassedCorrectly - 8]
{
 "args": {
  "bye": "false",
  "hi": "true"
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any?hi=true&bye=false"
}
---

[TestQueryParamsArePassedCorrectly - 9]
{
 "args": {
  "bye": "456.789",
  "hi": "123"
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any?hi=123&bye=456.789"
}
---

[TestQueryParamsArePassedCorrectly - 10]
{
 "args": {
  "bar": " space ",
  "bye": "true",
  "foo": "null",
  "hi": "1"
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any?hi=1&bye=true&foo=null&bar=%20space%20"
}
---

[TestQueryParamsArePassedCorrectly - 11]
{
 "args": {
  "bye": "goodbye/world",
  "hi": "hello world"
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any?hi=hello%20world&bye=goodbye%2Fworld"
}
---

[TestNonTerminatedPaths - 1]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "Referer": "<Any value>",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestRetainHostHeader - 1]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "testns1.com",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://testns1.com/any"
}
---

[TestOnlyGetAllowed - 2]
[]uint8{0x4d, 0x65, 0x74, 0x68, 0x6f, 0x64, 0x20, 0x4e, 0x6f, 0x74, 0x20, 0x41, 0x6c, 0x6c, 0x6f, 0x77, 0x65, 0x64, 0xa}
---

[TestNonTerminatedPaths - 2]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestNonTerminatedPaths - 3]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestAnyMethod - 1]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestAnyMethod - 2]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Content-Length": "0",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "POST",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestAnyMethod - 3]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Content-Length": "0",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "PUT",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestAnyMethod - 4]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Content-Length": "0",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "PATCH",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestAnyMethod - 5]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "DELETE",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestAnyMethod - 6]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "CONNECT",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestAnyMethod - 7]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "OPTIONS",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestAnyMethod - 8]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "TRACE",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestTerminatedPaths - 1]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "Referer": "<Any value>",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestTerminatedPaths - 2]
[]uint8{0x34, 0x30, 0x34, 0x20, 0x70, 0x61, 0x67, 0x65, 0x20, 0x6e, 0x6f, 0x74, 0x20, 0x66, 0x6f, 0x75, 0x6e, 0x64, 0xa}
---

[TestRedirectsNotFollowed - 1]
&http.Response{
    Status:     "302 Found",
    StatusCode: 302,
    Proto:      "HTTP/1.1",
    ProtoMajor: 1,
    ProtoMinor: 1,
    Header:     {
        "Content-Length": {"0"},
        "Location":       {"https://google.com"},
    },
    Body:             http.noBody{},
    ContentLength:    0,
    TransferEncoding: nil,
    Close:            false,
    Uncompressed:     false,
    Trailer:          {},
    Request:          &http.Request{
        Method: "GET",
        URL:    &url.URL{
            Scheme:      "http",
            Opaque:      "",
            User:        (*url.Userinfo)(nil),
            Host:        "<replaced>",
            Path:        "/testns1/httpbun/redirect-to",
            RawPath:     "",
            OmitHost:    false,
            ForceQuery:  false,
            RawQuery:    "url=https%3A%2F%2Fgoogle.com",
            Fragment:    "",
            RawFragment: "",
        },
        Proto:      "HTTP/1.1",
        ProtoMajor: 1,
        ProtoMinor: 1,
        Header:     {
        },
        Body:             nil,
        GetBody:          func() (io.ReadCloser, error) {...},
        ContentLength:    0,
        TransferEncoding: nil,
        Close:            false,
        Host:             "<replaced>",
        Form:             {},
        PostForm:         {},
        MultipartForm:    (*multipart.Form)(nil),
        Trailer:          {},
        RemoteAddr:       "",
        RequestURI:       "",
        TLS:              (*tls.ConnectionState)(nil),
        Cancel:           0x0,
        Response:         (*http.Response)(nil),
        Pattern:          "",
        ctx:              context.backgroundCtx{},
        pat:              (*http.pattern)(nil),
        matches:          nil,
        otherValues:      {},
    },
    TLS: (*tls.ConnectionState)(nil),
}
---

[TestPassthrough - 2]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any/hihi/%2F"
}
---

[TestPassthrough - 1]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/get"
}
---

[TestOnlyGetAllowed - 1]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any"
}
---

[TestWildcardRewrite - 1]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any/a/huhh"
}
---

[TestWildcardRewrite - 2]
{
 "args": {
  "abc": "lol",
  "x": "b"
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any/a/huhh?abc=lol&x=b"
}
---

[TestWildcardRewrite - 3]
{
 "args": {
  "abc": "魚",
  "x": "は"
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any/a/huhh?abc=魚&x=は"
}
---

[TestWildcardRewrite - 4]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any/slash%2Fshould-bekept/next"
}
---

[TestPathRewrite - 1]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any/a/efgh"
}
---

[TestPathRewrite - 2]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any/a/huhh"
}
---

[TestPathRewrite - 3]
{
 "args": {
  "abc": "lol",
  "x": "b"
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any/a/huhh?abc=lol&x=b"
}
---

[TestPathRewrite - 4]
{
 "args": {
  "abc": "魚",
  "x": "は"
 },
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any/a/huhh?abc=魚&x=は"
}
---

[TestPathRewrite - 5]
{
 "args": {},
 "data": "",
 "files": {},
 "form": {},
 "headers": {
  "Accept-Encoding": "gzip",
  "Host": "localhost:8080",
  "User-Agent": "Go-http-client/1.1"
 },
 "json": null,
 "method": "GET",
 "origin": "172.18.0.1",
 "url": "http://localhost:8080/any/slash%2Fshould-bekept/next"
}
---
